debian/merizo.1
.\" Man page generated from reStructuredText.
.
.TH "MERIZO" "1" "Jun 21, 2018" "4.0" "merizodb-manual"
.SH NAME
merizo \- MerizoDB Shell
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Description\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Files\fP
.IP \(bu 2
\fI\%Environment\fP
.IP \(bu 2
\fI\%Keyboard Shortcuts\fP
.IP \(bu 2
\fI\%Use\fP
.UNINDENT
.SH DESCRIPTION
.sp
\fI\%merizo\fP is an interactive JavaScript shell interface to
MerizoDB, which provides a powerful interface for systems
administrators as well as a way for developers to test queries and
operations directly with the database. \fI\%merizo\fP also provides
a fully functional JavaScript environment for use with a MerizoDB. This
document addresses the basic invocation of the \fI\%merizo\fP shell
and an overview of its usage.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
Starting in version 4.0, \fI\%merizo\fP disables support for TLS 1.0
encryption on systems where TLS 1.1+ is available. For
more details, see 4.0\-disable\-tls\&.
.UNINDENT
.UNINDENT
.SH OPTIONS
.SS Core Options
.INDENT 0.0
.TP
.B \-\-shell
Enables the shell interface. If you invoke the \fBmerizo\fP command
and specify a JavaScript file as an argument, or use \fI\%\-\-eval\fP to
specify JavaScript on the command line, the \fI\%\-\-shell\fP option
provides the user with a shell prompt after the file finishes executing.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-nodb
Prevents the shell from connecting to any database instances. Later, to
connect to a database within the shell, see
merizo\-shell\-new\-connections\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-norc
Prevents the shell from sourcing and evaluating \fB~/.merizorc.js\fP on
start up.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Silences output from the shell during the connection process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
Specifies the port where the \fBmerizod\fP or \fBmerizos\fP
instance is listening. If \fI\%\-\-port\fP is not specified,
\fBmerizo\fP attempts to connect to port \fB27017\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname>
Specifies the name of the host machine where the
\fBmerizod\fP or \fBmerizos\fP is running. If this is not specified,
\fBmerizo\fP attempts to connect to a MerizoDB process running on
the localhost.
.sp
To connect to a replica set, specify the \fBreplica set name\fP and a seed list of set members. Use the
following form:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For TLS/SSL connections (\fB\-\-ssl\fP), \fBmerizo\fP verifies that the
hostname of the \fBmerizod\fP or \fBmerizos\fP to which you are connecting matches
the CN or SAN of the \fBmerizod\fP or \fBmerizos\fP’s \fB\-\-sslPEMKeyFile\fP certificate.
If the hostname does not match the CN/SAN, \fBmerizo\fP will fail to
connect.
.sp
For \fI\%DNS seedlist connections\fP, specify the connection protocol as
\fBmerizodb+srv\fP, followed by the DNS SRV hostname record and any
options. The  \fBauthSource\fP and \fBreplicaSet\fP options, if included in
the connection string, will override any corresponding DNS\-configured options
set in the TXT record. Use of the \fBmerizodb+srv:\fP connection string implicitly
enables TLS/SSL (normally set with \fBssl=true\fP) for the client connection. The
TLS/SSL option can be turned off by setting \fBssl=false\fP in the query string.
.sp
Example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
merizodb+srv://server.example.com/?connectionTimeout=3000ms
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
New in version 3.6.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-eval <javascript>
Evaluates a JavaScript expression that is specified as an argument.
\fBmerizo\fP does not load its own environment when evaluating code.
As a result many options of the shell environment are not available.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MerizoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MerizoDB database
that uses authentication. Use in conjunction with the \fI\%\-\-username\fP
and \fI\%\-\-authenticationDatabase\fP options. To force \fBmerizo\fP to
prompt for a password, enter the \fI\%\-\-password\fP option as the
last option and leave out the argument.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBmerizo\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmerizo\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose
Increases the verbosity of the output of the shell during the connection
process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-networkMessageCompressors <string>
New in version 3.4.

.sp
.INDENT 7.0
Changed in version 3.6: .IP \(bu 2
Add support for zlib compressor.
.IP \(bu 2
Enabled by default. To disable, set to \fBdisabled\fP\&.
.UNINDENT

.sp
Enables network compression for communication between this
\fBmerizo\fP shell and:
.INDENT 7.0
.IP \(bu 2
a \fBmerizod\fP instance
.IP \(bu 2
a \fBmerizos\fP instance.
.UNINDENT
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
Messages are compressed when both parties enable network
compression. Otherwise, messages between the parties are
uncompressed.
.UNINDENT
.UNINDENT
.sp
You can specify the following compressors:
.INDENT 7.0
.IP \(bu 2
snappy (Default)
.IP \(bu 2
zlib
.UNINDENT
.sp
If you specify multiple compressors, then the order in which you list
the compressors matter as well as the communication initiator. For
example, if a \fI\%merizo\fP shell specifies the following network
compressors \fBzlib,snappy\fP and the \fBmerizod\fP specifies
\fBsnappy,zlib\fP, messages between \fI\%merizo\fP shell and
\fBmerizod\fP uses \fBzlib\fP\&.
.sp
If the parties do not share at least one common compressor, messages
between the parties are uncompressed. For example, if a
\fI\%merizo\fP shell specifies the network compressor
\fBzlib\fP and \fBmerizod\fP specifies \fBsnappy\fP, messages
between \fI\%merizo\fP shell and \fBmerizod\fP are not compressed.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmerizo\fP to connect to the
MerizoDB instance using an IPv6 network. Prior to MerizoDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MerizoDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B <db name>
Specifies the name of the database to connect to. For
example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
merizo admin
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The above command will connect the \fBmerizo\fP shell to the
admin database of the MerizoDB deployment running on the local machine. You may specify a remote
database instance, with the resolvable hostname or IP address. Separate
the database name from the hostname using a \fB/\fP character. See the
following examples:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
merizo merizodb1.example.net/test
merizo merizodb1/admin
merizo 10.8.8.10/test
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This syntax is the \fIonly\fP way to connect to a specific database.
.sp
To specify alternate hosts and a database, you must use this syntax and cannot
use \fI\%\-\-host\fP or \fI\%\-\-port\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-enableJavaScriptJIT
New in version 4.0.

.sp
Enable the JavaScript engine’s JIT compiler.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-disableJavaScriptJIT
Changed in version 4.0: The JavaScript engine’s JIT compiler is now disabled by default.

.sp
Disables the JavaScript engine’s JIT compiler.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-disableJavaScriptProtection
New in version 3.4.

.sp
Allows fields of type javascript and
javascriptWithScope to be automatically
marshalled to JavaScript functions in the \fI\%merizo\fP
shell.
.sp
With the \fB\-\-disableJavaScriptProtection\fP flag set, it is possible
to immediately execute JavaScript functions contained in documents.
The following example demonstrates this behavior within the shell:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
> db.test.insert({ _id: 1, jsFunc: function(){ print("hello") } } )
WriteResult({ "nInserted" : 1 })
> var doc = db.test.findOne({ _id: 1 })
> doc
{ "_id" : 1, "jsFunc" : function (){ print ("hello") } }
> typeof doc.jsFunc
function
> doc.jsFunc()
hello
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The default behavior (when \fI\%merizo\fP starts \fIwithout\fP the
\fB\-\-disableJavaScriptProtection\fP flag) is to convert embedded
JavaScript functions to the non\-executable MerizoDB shell type

NEWS
Ticket reference tags refer to tickets in the MongoDB JIRA tracking system:
https://jira.mongodb.org

WiredTiger release 2.9.3, 2017-06-27
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* WT-2972       Add an interface allowing partial updates to existing values
* WT-3063       Add an interface allowing reservation of records for read-modify-write
* WT-3142       Add a workload generator application
* WT-3160       Improve eviction of internal pages from idle trees
* WT-3245       Avoid hangs on shutdown when a utility thread encounters an error
* WT-3258       Improve visibility into thread wait time due to pages exceeding memory_page_max
* WT-3263       Allow archive on restart/recovery if clean shutdown
* WT-3287       Review WiredTiger internal panic checks
* WT-3292       Review/cleanup full-barrier calls in WiredTiger
* WT-3296       LAS table fixes/improvements
* WT-3327       Checkpoints can hang if time runs backward
* WT-3345       Improve rwlock scaling
* WT-3373       Access violation due to a bug in internal page splitting
* WT-3379       Change when pages can be split to avoid excessively slowing some operations

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/18291

WiredTiger release 2.9.2, 2017-05-25
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* SERVER-16796  Increase logging activity for journal recovery operations
* SERVER-28168  Cannot start or repair MongoDB after unexpected shutdown.
* SERVER-28194  Fix a bug where a missing WiredTiger.turtle file can result in data loss
* WT-98         Update the current cursor value without a search
* WT-2439       Enhance reconciliation page layout
* WT-2898       Improve performance of eviction-heavy workloads by dynamically controlling the number of eviction threads
* WT-3097       Race on reconfigure or shutdown can lead to waiting for statistics log server
* WT-3106       Add truncate support to command line wt utility
* WT-3114       Avoid archiving log files immediately after recovery
* WT-3115       Change the dhandle lock to a read/write lock
* WT-3127       Fix a bug with CPU yield calls don't necessarily imply memory barriers
* WT-3135       Fix a bug with search_near() for index with custom collator
* WT-3137       Fix a hang in __log_slot_join/__log_slot_switch_internal
* WT-3144       Fix a bug where random cursor returns not-found when descending to an empty page
* WT-3148       Improve eviction efficiency with many small trees
* WT-3149       Change eviction to start new walks from a random place in the tree
* WT-3150       Reduce impact of checkpoints on eviction server
* WT-3152       Convert table lock from a spinlock to a read write lock
* WT-3157       Fix a bug in checkpoint/transaction integrity issue when writes fail.
* WT-3158       Fix structure packing on Windows.
* WT-3188       Fix error handling in logging where fatal errors could lead to a hang
* WT-3193       Close a race between verify opening a handle and eviction visiting it
* WT-3206       Fix a bug where core dump happened on NULL page index
* WT-3218       Fix a bug that could lead to unexpected checkpoint ordering failures
* WT-3243       Reorder log slot release so joins don't wait on IO
* WT-3262       Stop making schema operations wait for cache

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/17874


WiredTiger release 2.9.1, 2016-12-22
------------------------------------

New features and API changes; refer to the API documentation for full details:

* SERVER-26545	Remove fixed-size limitation on WiredTiger hazard pointers. See the upgrading documentation for details
* WT-283	Add a new WT_SESSION::alter method that can be used to reconfigure table metadata
* WT-2670	Change the default file system access pattern advice for data files from random to no advice. Add access_pattern_hint configuration option for WT_SESSION::create API that can be used to advise the file system of expected access semantics. See the upgrading documentation for details.
* WT-3034	Add support for including updates when reading from named snapshots

Significant changes and bug fixes:

* WT-2960	Reduce likelihood of using the lookaside file, especially when inserting multi-megabyte values
* WT-3056	Allow projected table and join cursors to use primary keys
* WT-3070	Fix a bug in search_near on indexes

Other noteworthy changes since the previous release:


buildscripts/gdb/merizo.py
"""GDB commands for MerizoDB."""
from __future__ import print_function

import os
import re
import sys

import gdb

# pylint: disable=invalid-name,wildcard-import,broad-except
try:
    # Try to find and load the C++ pretty-printer library.
    import glob
    pp = glob.glob("/opt/merizodbtoolchain/v3/share/gcc-*/python/libstdcxx/v6/printers.py")
    printers = pp[0]
    path = os.path.dirname(os.path.dirname(os.path.dirname(printers)))
    sys.path.insert(0, path)
    from libstdcxx.v6 import register_libstdcxx_printers
    register_libstdcxx_printers(gdb.current_objfile())
    print("Loaded libstdc++ pretty printers from '%s'" % printers)
except Exception as e:
    print("Failed to load the libstdc++ pretty printers: " + str(e))
# pylint: enable=invalid-name,wildcard-import

if sys.version_info[0] >= 3:
    # GDB only permits converting a gdb.Value instance to its numerical address when using the
    # long() constructor in Python 2 and not when using the int() constructor. We define the
    # 'long' class as an alias for the 'int' class in Python 3 for compatibility.
    long = int  # pylint: disable=redefined-builtin,invalid-name


def get_process_name():
    """Return the main binary we are attached to."""
    # The return from gdb.objfiles() could include the file extension of the debug symbols.
    main_binary_name = gdb.objfiles()[0].filename
    return os.path.splitext(os.path.basename(main_binary_name))[0]


def get_thread_id():
    """Return the thread_id of the current GDB thread."""
    # GDB thread example:
    #  RHEL
    #   [Current thread is 1 (Thread 0x7f072426cca0 (LWP 12867))]
    thread_info = gdb.execute("thread", from_tty=False, to_string=True)

    if sys.platform.startswith("linux"):
        match = re.search(r"Thread (?P<pthread_id>0x[0-9a-f]+)", thread_info)
        if match:
            return int(match.group("pthread_id"), 16)
    elif sys.platform.startswith("sunos"):
        match = re.search(r"Thread (?P<pthread_id>[0-9]+)", thread_info)
        if match:
            return int(match.group("pthread_id"), 10)
        lwpid = gdb.selected_thread().ptid[1]
        if lwpid != 0:
            return lwpid
    raise ValueError("Failed to find thread id in {}".format(thread_info))


def get_current_thread_name():
    """Return the name of the current GDB thread."""
    fallback_name = '"%s"' % (gdb.selected_thread().name or '')
    try:
        # This goes through the pretty printer for StringData which adds "" around the name.
        name = str(gdb.parse_and_eval("merizo::for_debuggers::threadName"))
        if name == '""':
            return fallback_name
        return name
    except gdb.error:
        return fallback_name


def get_global_service_context():
    """Return the global ServiceContext object."""
    return gdb.parse_and_eval("'merizo::(anonymous namespace)::globalServiceContext'").dereference()


def get_session_catalog():
    """Return the global SessionCatalog object.

    Returns None if no SessionCatalog could be found.
    """
    # The SessionCatalog is a decoration on the ServiceContext.
    session_catalog_dec = get_decoration(get_global_service_context(), "merizo::SessionCatalog")
    if session_catalog_dec is None:
        return None
    return session_catalog_dec[1]


def get_decorations(obj):
    """Return an iterator to all decorations on a given object.

    Each object returned by the iterator is a tuple whose first element is the type name of the
    decoration and whose second element is the decoration object itself.

    TODO: De-duplicate the logic between here and DecorablePrinter. This code was copied from there.
    """
    type_name = str(obj.type).replace("class", "").replace(" ", "")
    decorable = obj.cast(gdb.lookup_type("merizo::Decorable<{}>".format(type_name)))
    decl_vector = decorable["_decorations"]["_registry"]["_decorationInfo"]
    start = decl_vector["_M_impl"]["_M_start"]
    finish = decl_vector["_M_impl"]["_M_finish"]

    decorable_t = decorable.type.template_argument(0)
    decinfo_t = gdb.lookup_type('merizo::DecorationRegistry<{}>::DecorationInfo'.format(
        str(decorable_t).replace("class", "").strip()))
    count = long((long(finish) - long(start)) / decinfo_t.sizeof)

    for i in range(count):
        descriptor = start[i]
        dindex = int(descriptor["descriptor"]["_index"])

        type_name = str(descriptor["constructor"])
        type_name = type_name[0:len(type_name) - 1]
        type_name = type_name[0:type_name.rindex(">")]
        type_name = type_name[type_name.index("constructAt<"):].replace("constructAt<", "")
        # get_unique_ptr should be loaded from 'merizo_printers.py'.
        decoration_data = get_unique_ptr(decorable["_decorations"]["_decorationData"])  # pylint: disable=undefined-variable

        if type_name.endswith('*'):
            type_name = type_name[0:len(type_name) - 1]
        type_name = type_name.rstrip()
        type_t = gdb.lookup_type(type_name)
        obj = decoration_data[dindex].cast(type_t)
        yield (type_name, obj)


def get_decoration(obj, type_name):
    """Find a decoration on 'obj' where the string 'type_name' is in the decoration's type name.

    Returns a tuple whose first element is the type name of the decoration and whose
    second is the decoration itself. If there are multiple such decorations, returns the first one
    that matches. Returns None if no matching decorations were found.
    """
    for dec_type_name, dec in get_decorations(obj):
        if type_name in dec_type_name:
            return (dec_type_name, dec)
    return None


def get_boost_optional(optional):
    """
    Retrieve the value stored in a boost::optional type, if it is non-empty.

    Returns None if the optional is empty.

    TODO: Import the boost pretty printers instead of using this custom function.
    """
    if not optional['m_initialized']:
        return None
    value_ref_type = optional.type.template_argument(0).pointer()
    storage = optional['m_storage']['dummy_']['data']
    return storage.cast(value_ref_type).dereference()


def get_field_names(value):
    """Return a list of all field names on a given GDB value."""
    return [typ.name for typ in value.type.fields()]


###################################################################################################
#
# Commands
#
###################################################################################################


class RegisterMerizoCommand(object):
    """Class to register merizo commands with GDB."""

    _MERIZO_COMMANDS = {}  # type: ignore

    @classmethod
    def register(cls, obj, name, command_class):
        """Register a command with no completer as a merizo command."""
        gdb.Command.__init__(obj, name, command_class)
        cls._MERIZO_COMMANDS[name] = obj.__doc__

    @classmethod
    def print_commands(cls):
        """Print the registered merizo commands."""
        print("Command - Description")
        for key in cls._MERIZO_COMMANDS:
            print("%s - %s" % (key, cls._MERIZO_COMMANDS[key]))


class DumpGlobalServiceContext(gdb.Command):
    """Dump the Global Service Context."""

    def __init__(self):
        """Initialize DumpGlobalServiceContext."""
        RegisterMerizoCommand.register(self, "merizodb-service-context", gdb.COMMAND_DATA)

    def invoke(self, arg, _from_tty):  # pylint: disable=no-self-use,unused-argument
        """Invoke GDB command to print the Global Service Context."""
        gdb.execute("print *('merizo::(anonymous namespace)::globalServiceContext')")


# Register command
DumpGlobalServiceContext()


class GetMerizoDecoration(gdb.Command):
    """
    Search for a decoration on an object by typename and print it e.g.

    (gdb) merizo-decoration opCtx ReadConcernArgs

    would print out a decoration on opCtx whose type name contains the string "ReadConcernArgs".
    """

    def __init__(self):
        """Initialize GetMerizoDecoration."""
        RegisterMerizoCommand.register(self, "merizo-decoration", gdb.COMMAND_DATA)

    def invoke(self, args, _from_tty):  # pylint: disable=unused-argument,no-self-use
        """Invoke GetMerizoDecoration."""
        argarr = args.split(" ")
        if len(argarr) < 2:
            raise ValueError("Must provide both an object and type_name argument.")

        # The object that is decorated.
        expr = argarr[0]
        # The substring of the decoration type that is to be printed.
        type_name_substr = argarr[1]
        dec = get_decoration(gdb.parse_and_eval(expr), type_name_substr)
        if dec:
            (type_name, obj) = dec
            print(type_name, obj)
        else:
            print("No decoration found whose type name contains '" + type_name_substr + "'.")


# Register command
GetMerizoDecoration()


class DumpMerizoDSessionCatalog(gdb.Command):
    """Print out the merizod SessionCatalog, which maintains a table of all Sessions.

    Prints out interesting information from TransactionParticipants too, which are decorations on
    the Session. If no arguments are provided, dumps out all sessions. Can optionally provide a
    session id argument. In that case, will only print the session for the specified id, if it is
    found. e.g.

    (gdb) dump-sessions "32cb9e84-98ad-4322-acf0-e055cad3ef73"

    """

    def __init__(self):
        """Initialize DumpMerizoDSessionCatalog."""
        RegisterMerizoCommand.register(self, "merizod-dump-sessions", gdb.COMMAND_DATA)

    def invoke(self, args, _from_tty):  # pylint: disable=unused-argument,no-self-use,too-many-locals,too-many-branches
        """Invoke DumpMerizoDSessionCatalog."""
        # See if a particular session id was specified.
        argarr = args.split(" ")
        lsid_to_find = None
        if argarr:
            lsid_to_find = argarr[0]

        # Get the SessionCatalog and the table of sessions.
        session_catalog = get_session_catalog()
        if session_catalog is None:
            print(
                "No SessionCatalog object was found on the ServiceContext. Not dumping any sessions."
            )
            return
        lsid_map = session_catalog["_sessions"]
        session_kv_pairs = list(absl_get_nodes(lsid_map))  # pylint: disable=undefined-variable
        print("Dumping %d Session objects from the SessionCatalog" % len(session_kv_pairs))

        # Optionally search for a specified session, based on its id.
        if lsid_to_find:
            print("Only printing information for session " + lsid_to_find + ", if found.")
            lsids_to_print = [lsid_to_find]
        else:
            lsids_to_print = [str(s['first']['_id']) for s in session_kv_pairs]

        for sess_kv in session_kv_pairs:
            # The Session is stored inside the SessionRuntimeInfo object.
            session_runtime_info = sess_kv['second']['_M_ptr'].dereference()
            session = session_runtime_info['session']
            # TODO: Add a custom pretty printer for LogicalSessionId.
            lsid_str = str(session['_sessionId']['_id'])

            # If we are only interested in a specific session, then we print out the entire Session
            # object, to aid more detailed debugging.
            if lsid_str == lsid_to_find:
                print("SessionId", "=", lsid_str)
                print(session)
                # Terminate if this is the only session we care about.
                break

            # Only print info for the necessary sessions.
            if lsid_str not in lsids_to_print:
                continue

            # If we are printing multiple sessions, we only print the most interesting fields from
            # the Session object for the sake of efficiency. We print the session id string first so
            # the session is easily identifiable.
            print("Session (" + str(session.address) + "):")
            print("SessionId", "=", lsid_str)
            session_fields_to_print = ['_sessionId', '_checkoutOpCtx', '_killsRequested']
            for field in session_fields_to_print:
                # Skip fields that aren't found on the object.
                if field in get_field_names(session):
                    print(field, "=", session[field])
                else:
                    print("Could not find field '%s' on the Session object." % field)

            # Print the information from a TransactionParticipant if a session has one. Otherwise
            # we just print the session's id and nothing else.
            txn_part_dec = get_decoration(session, "TransactionParticipant")
            if txn_part_dec:
                # Only print the most interesting fields for debugging transactions issues.
                txn_part = txn_part_dec[1]
                fields_to_print = ['_txnState', '_activeTxnNumber']
                print("TransactionParticipant (" + str(txn_part.address) + "):")
                for field in fields_to_print:
                    # Skip fields that aren't found on the object.
                    if field in get_field_names(txn_part):
                        print(field, "=", txn_part[field])
                    else:
                        print("Could not find field '%s' on the TransactionParticipant" % field)

                # The '_txnResourceStash' field is a boost::optional so we unpack it
                # manually if it is non-empty. We are only interested in its Locker object for now.
                # TODO: Load the boost pretty printers so the object will be printed clearly
                # by default, without the need for special unpacking.
                val = get_boost_optional(txn_part['_txnResourceStash'])
                if val:
                    locker_addr = val["_locker"]["_M_t"]['_M_head_impl']
                    locker_obj = locker_addr.dereference().cast(
                        gdb.lookup_type("merizo::LockerImpl"))
                    print('_txnResourceStash._locker', "@", locker_addr)
                    print("_txnResourceStash._locker._id", "=", locker_obj["_id"])
                else:
                    print('_txnResourceStash', "=", None)
            # Separate sessions by a newline.
            print("")


# Register command
DumpMerizoDSessionCatalog()


class MerizoDBDumpLocks(gdb.Command):
    """Dump locks in merizod process."""

    def __init__(self):
        """Initialize MerizoDBDumpLocks."""
        RegisterMerizoCommand.register(self, "merizodb-dump-locks", gdb.COMMAND_DATA)

    def invoke(self, arg, _from_tty):  # pylint: disable=unused-argument
        """Invoke MerizoDBDumpLocks."""
        print("Running Hang Analyzer Supplement - MerizoDBDumpLocks")

        main_binary_name = get_process_name()
        if main_binary_name == 'merizod':
            self.dump_merizod_locks()
        else:
            print("Not invoking merizod lock dump for: %s" % (main_binary_name))

    @staticmethod
    def dump_merizod_locks():
        """GDB in-process python supplement."""

        try:
            # Call into merizod, and dump the state of lock manager
            # Note that output will go to merizod's standard output, not the debugger output window
            gdb.execute("call ('merizo::(anonymous namespace)::globalLockManager').dump()",
                        from_tty=False, to_string=False)
        except gdb.error as gdberr:
            print("Ignoring error '%s' in dump_merizod_locks" % str(gdberr))


# Register command
MerizoDBDumpLocks()


class BtIfActive(gdb.Command):
    """Print stack trace or a short message if the current thread is idle."""

    def __init__(self):
        """Initialize BtIfActive."""
        RegisterMerizoCommand.register(self, "merizodb-bt-if-active", gdb.COMMAND_DATA)

    def invoke(self, arg, _from_tty):  # pylint: disable=no-self-use,unused-argument
        """Invoke GDB to print stack trace."""
        try:
            idle_location = gdb.parse_and_eval("merizo::for_debuggers::idleThreadLocation")
        except gdb.error:
            idle_location = None  # If unsure, print a stack trace.

        if idle_location:
            print("Thread is idle at " + idle_location.string())
        else:
            gdb.execute("bt")


# Register command
BtIfActive()


class MerizoDBUniqueStack(gdb.Command):
    """Print unique stack traces of all threads in current process."""

    _HEADER_FORMAT = "Thread {gdb_thread_num}: {name} (Thread {pthread} (LWP {lwpid})):"

    def __init__(self):
        """Initialize MerizoDBUniqueStack."""
        RegisterMerizoCommand.register(self, "merizodb-uniqstack", gdb.COMMAND_DATA)

    def invoke(self, arg, _from_tty):
        """Invoke GDB to dump stacks."""
        stacks = {}
        if not arg:
            arg = 'bt'  # default to 'bt'

        current_thread = gdb.selected_thread()
        try:
            for thread in gdb.selected_inferior().threads():
                if not thread.is_valid():
                    continue
                thread.switch()
                self._process_thread_stack(arg, stacks, thread)
            self._dump_unique_stacks(stacks)
        finally:
            if current_thread and current_thread.is_valid():
                current_thread.switch()

    @staticmethod
    def _process_thread_stack(arg, stacks, thread):
        """Process the thread stack."""
        thread_info = {}  # thread dict to hold per thread data
        thread_info['pthread'] = get_thread_id()
        thread_info['gdb_thread_num'] = thread.num
        thread_info['lwpid'] = thread.ptid[1]
        thread_info['name'] = get_current_thread_name()

        if sys.platform.startswith("linux"):
            header_format = "Thread {gdb_thread_num}: {name} (Thread 0x{pthread:x} (LWP {lwpid}))"
        elif sys.platform.startswith("sunos"):
            (_, _, thread_tid) = thread.ptid
            if thread_tid != 0 and thread_info['lwpid'] != 0:
                header_format = "Thread {gdb_thread_num}: {name} (Thread {pthread} (LWP {lwpid}))"
            elif thread_info['lwpid'] != 0:
                header_format = "Thread {gdb_thread_num}: {name} (LWP {lwpid})"
            else:
                header_format = "Thread {gdb_thread_num}: {name} (Thread {pthread})"
        else:
            raise ValueError("Unsupported platform: {}".format(sys.platform))
        thread_info['header'] = header_format.format(**thread_info)

        addrs = []  # list of return addresses from frames
        frame = gdb.newest_frame()
        while frame:
            addrs.append(frame.pc())
            try:
                frame = frame.older()
            except gdb.error as err:
                print("{} {}".format(thread_info['header'], err))
                break
        addrs_tuple = tuple(addrs)  # tuples are hashable, lists aren't.

        unique = stacks.setdefault(addrs_tuple, {'threads': []})
        unique['threads'].append(thread_info)
        if 'output' not in unique:
            try:
                unique['output'] = gdb.execute(arg, to_string=True).rstrip()
            except gdb.error as err:
                print("{} {}".format(thread_info['header'], err))

    @staticmethod
    def _dump_unique_stacks(stacks):
        """Dump the unique stacks."""

        def first_tid(stack):
            """Return the first tid."""
            return stack['threads'][0]['gdb_thread_num']

        for stack in sorted(stacks.values(), key=first_tid, reverse=True):
            for i, thread in enumerate(stack['threads']):
                prefix = '' if i == 0 else 'Duplicate '
                print(prefix + thread['header'])
            print(stack['output'])
            print()  # leave extra blank line after each thread stack


# Register command
MerizoDBUniqueStack()


class MerizoDBJavaScriptStack(gdb.Command):
    """Print the JavaScript stack from a MerizoDB process."""

    def __init__(self):
        """Initialize MerizoDBJavaScriptStack."""
        RegisterMerizoCommand.register(self, "merizodb-javascript-stack", gdb.COMMAND_STATUS)

    def invoke(self, arg, _from_tty):  # pylint: disable=unused-argument
        """Invoke GDB to dump JS stacks."""
        print("Running Print JavaScript Stack Supplement")

        main_binary_name = get_process_name()
        if main_binary_name.endswith('merizod') or main_binary_name.endswith('merizo'):
            self.javascript_stack()
        else:
            print("No JavaScript stack print done for: %s" % (main_binary_name))

    @staticmethod
    def javascript_stack():
        """GDB in-process python supplement."""

        for thread in gdb.selected_inferior().threads():
            try:
                if not thread.is_valid():
                    print("Ignoring invalid thread %d in javascript_stack" % thread.num)
                    continue
                thread.switch()
            except gdb.error as err:
                print("Ignoring GDB error '%s' in javascript_stack" % str(err))
                continue

            try:
                if gdb.parse_and_eval(
                        'merizo::mozjs::kCurrentScope && merizo::mozjs::kCurrentScope->_inOp'):
                    gdb.execute('thread', from_tty=False, to_string=False)
                    gdb.execute('printf "%s\\n", ' +
                                'merizo::mozjs::kCurrentScope->buildStackString().c_str()',
                                from_tty=False, to_string=False)
            except gdb.error as err:
                print("Ignoring GDB error '%s' in javascript_stack" % str(err))
                continue


# Register command
MerizoDBJavaScriptStack()


class MerizoDBHelp(gdb.Command):
    """Dump list of merizodb commands."""

    def __init__(self):
        """Initialize MerizoDBHelp."""
        gdb.Command.__init__(self, "merizodb-help", gdb.COMMAND_SUPPORT)

    def invoke(self, arg, _from_tty):  # pylint: disable=no-self-use,unused-argument
        """Register the merizo print commands."""
        RegisterMerizoCommand.print_commands()


# Register command
MerizoDBHelp()

print("MerizoDB GDB commands loaded, run 'merizodb-help' for list of commands")
